### PR TITLE
Refactor of NetworkBase

### PR DESCRIPTION
**Pull request by sesas**
@gimballock @odandia: Hey guys, check this PR out

### CODE REVIEW

File: src/examples/basic.py [hunk 1]
```diff
@@ -9,15 +7,8 @@ class MyEnvClass(NetworkBase):
...
-        self.add_child_template(Bastion())
-
-        # Add a preconfigured ASG + ELB as a child stack in the environment
-        self.add_child_template(HaCluster(
-            name="MyCluster",
-            min_size=2, max_size=3,
-            instance_type='t2.micro'))
-
+        # Do custom troposphere resource creation here
+        super(MyEnvClass, self).create_hook()
```
Comment (gimballock) [inline src/examples/basic.py:11]: At this point the example no longer does anything.  What's the rationale here?

Comment (sesas) [inline src/examples/basic.py:11]: TBH the client I am working w/ had already removed that code.  You should look at the history of this file too, it went from barebone to this.  Probably it's worth it to keep both examples: barebone +  bastion&ha-nat

**Comment by sesas** [comment]
I found some bugs in this PR. let me test it a bit more and I'll push some more changes in a bit.

**Comment by sesas** [comment]
This is part of the issue I am running into: https://github.com/cloudtools/troposphere/issues/379

I'm outputting a `FindInMap` function, but troposphere is raising an error.

**Comment by sesas** [comment]
I need to do the subnet hook up too...

File: src/environmentbase/networkbase.py [hunk 2]
```diff
@@ -347,11 +364,59 @@ def add_vpn_gateway(self,
...
+
+        base_network_template = BaseNetwork('BaseNetwork', network_config, boto_config, nat_config, az_count)
+        self.add_child_template(base_network_template)
+
+        # TODO: make outputs passing less fragile:
+        # base_network_template.build_hook()
+        # for output in base_network_template.outputs:
+        #     self.template.add_output(Output('', Value=GetAtt(base_network_template.name, '')))
+
+        self.template.add_output(Output('commonSecurityGroupId',
```
Comment (gimballock) [inline src/environmentbase/networkbase.py]: These outputs can be moved inside of BaseNetwork

File: src/environmentbase/networkbase.py [hunk 3]
```diff
@@ -347,11 +364,59 @@ def add_vpn_gateway(self,
...
-        gateway_connection = self.template.add_resource(ec2.VPCGatewayAttachment('vpnGatewayAttachment',
-            VpcId=self.template.vpc_id,
-            InternetGatewayId=self.template.igw,
+        gateway_connection = self.add_resource(ec2.VPCGatewayAttachment('vpnGatewayAttachment',
+            VpcId=self.vpc_id,
+            InternetGatewayId=self.igw,
             VpnGatewayId=gateway))
+
+
+class NetworkBase(EnvironmentBase):
```
Comment (gimballock) [inline src/environmentbase/networkbase.py:391]: Ultimately NetworkBase class should not be necessary.  BaseNetwork class should be moved to a new file inside the /patterns directory and this file 'network.py' should be removed. But these tasks can be done in a follow on commit.

File: src/environmentbase/networkbase.py [hunk 4]
```diff
@@ -63,24 +89,14 @@ def construct_network(self):
...
-        self.initialize_template()
-        self.construct_network()
-        self.create_hook()
-        self.serialize_templates()
+            self._azs.append(FindInMap('RegionMap', Ref('AWS::Region'), 'az' + str(x) + 'Name'))
 
     def add_vpc_az_mapping(self,
                            boto_config,
-                           az_count=2):
+                           az_count=2, cached=False):
```
Comment (gimballock) [inline src/environmentbase/networkbase.py:101]: Can you add a \@param comment for cached parameter?

File: src/environmentbase/networkbase.py [hunk 5]
```diff
@@ -13,35 +13,61 @@
...
-
-class NetworkBase(EnvironmentBase):
-    """
-    EnvironmentBase controller containing a root template with all of the base networking infrastructure
-    for a common deployment within AWS. This is intended to be the 'base' stack for deploying child stacks
-    """
+from template import Template
+
+
+AWS_MAPPING = dict(
```
Comment (gimballock) [inline src/environmentbase/networkbase.py:19]: Would like to externalize this at some point.

File: src/environmentbase/networkbase.py [hunk 6]
```diff
@@ -347,11 +364,59 @@ def add_vpn_gateway(self,
...
+
+
+class NetworkBase(EnvironmentBase):
+    """
+    EnvironmentBase controller containing a root template with all of the base networking infrastructure
+    for a common deployment within AWS. This is intended to be the 'base' stack for deploying child stacks
+    """
+
+    def create_hook(self):
+        network_config = self.config.get('network', {})
```
Comment (gimballock) [inline src/environmentbase/networkbase.py:400]: We should think about the structure of the config sections used here.  We can let BaseNetwork own its own config section so we don't need to pass it in.

File: src/environmentbase/networkbase.py (hunk 2 above)
Comment (sesas) [inline src/environmentbase/networkbase.py]: These outputs were in the original version of the root template, and they are in the `BaseNetwork` child stack as well.  I just wanted to maintain the same outputs and parameters for the root stack before and after the refactor.

File: src/environmentbase/networkbase.py (hunk 3 above)
Comment (sesas) [inline src/environmentbase/networkbase.py:391]: That's true, but I wanted to manage the migration from one version to the other.  I wanted to be backward compatible as much as possible. And that's also why I kept the `BaseNetwork` template in here too.

File: src/environmentbase/networkbase.py (hunk 4 above)
Comment (sesas) [inline src/environmentbase/networkbase.py:101]: True, I'll add it.

Comment (sesas) [inline src/environmentbase/networkbase.py:101]: I was also thinking of adding a new `config.json` parameter to toggle this feature on or off, but decided to keep the changes to a minimum for now.  I'll put a comment there about it, as a `TODO`

File: src/environmentbase/networkbase.py (hunk 5 above)
Comment (sesas) [inline src/environmentbase/networkbase.py:19]: True. I'll do it in a later incremental enhancement of env-base

Comment (sesas) [inline src/environmentbase/networkbase.py:19]: Probably I should add a comment for it though.

File: src/environmentbase/networkbase.py (hunk 6 above)
Comment (sesas) [inline src/environmentbase/networkbase.py:400]: We have to do add the config.json schema generation for this new template as well.

**Comment by sesas** [comment]
@gimballock new commits are in ^

File: src/environmentbase/networkbase.py [hunk 7]
```diff
@@ -347,11 +373,62 @@ def add_vpn_gateway(self,
...
+        #     self.template.add_output(Output('', Value=GetAtt(base_network_template.name, '')))
+
+        self.template._subnets = base_network_template._subnets
+
+        self.template.add_output(Output('commonSecurityGroupId', 
+                Value=GetAtt(base_network_template.name, 'Outputs.commonSecurityGroupId')))
+
+        self.template.add_output(Output('vpcId', 
+                Value=GetAtt(base_network_template.name, 'Outputs.vpcId')))
+        self.template._vpc_id = GetAtt(base_network_template.name, 'Outputs.vpcId')
```
Comment (sesas) [inline src/environmentbase/networkbase.py:411]: @gimballock we need to do this for the other python attributes, `._vpc_cidr`, ...

Comment (sesas) [inline src/environmentbase/networkbase.py:411]: `self._vpc_cidr =`
`self._vpc_id =`
`self._common_security_group =`
`self._utility_bucket =`
`self._igw =`

**Comment by sesas** [comment]
I suppose the branch got fixed after I created a new commit and pushed to it.

File: src/environmentbase/networkbase.py (hunk 4 above)
Comment (sesas) [inline src/environmentbase/networkbase.py:101]: irrelevant as of now because of one of the features you introduced later

**Comment by gimballock** [comment]
I reviewed this branch and it looks good to merge

**Pull request merged**